docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,197 @@
+# Support bin packing of fault domains
+
+## Metadata
+
+* Authors: @johscheuer
+* Created: 2021-10-01
+* Updated: 2021-10-01
+
+## Background
+
+In the current implementation of the operator we try to distribute our Pod across as many fault domains as possible.
+If a cluster has more fault domains than the operator creates Pods for a specific class all Pods will be on a different fault domain.
+This has the drawback that the operator can only interact on a single Pod for safe operations e.g. to minimize the risk of data loss.
+For clusters with a non-trivial amount of Pods this means that some operational tasks will take very long.
+A better way would be to allow the operator to bin pack Pods in the same fault domain.
+
+## General Design Goals
+
+* Provide a way how a user can specify the number of logical fault domains
+* The user should be able to decide if the number of logical fault domains is a requirement (default should be optional)
+
+## Current Implementation
+
+The current implementation adds a `preferredDuringSchedulingIgnoredDuringExecution` `PodAntiAffinity` to every Pod [code](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/internal/pod_models.go#L308-L334).
+The label selector will contain the process class and the cluster name, to ensure processes with the same class are distributed across different fault domains is possible.
+A user can define additional `PodAntiAffinity`s but can't prevent the operator from creating the default `PodAntiAffinity`.
+The only exception is the special fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+In the design we will use the following terms:
+
+- `physical fault domain`: This refers to a fault domain that is available in Kubernetes e.g. a machine running the Kubernetes worker or a higher level fault domain like a rack, data hall etc.
+  This will reflect an underlying physical fault domain.
+- `logical fault domain`: The logical fault domain is not something that refers directly to a physical fault domain and can also be spread across multiple physical fault domains.
+  The idea is to have a way to group some processes of a FoundationDB cluster together with the same zone ID, this zone ID will be a logical value like "zone-1".
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we will use a prefix for the `BinPack` mode and a value defining in which zone the process should be running.
+We will also set the `FDB_ZONE_ID` environment variable on the Pod level, the default value of this field is currently the hostname.
```

Comment:
Is this implying that we will set the FDB_ZONE_ID environment variable _to the distribution key_? That's my assumption, but it may be good to make that explicit if so.